Opencensus compatibility specification (#1332)
open-telemetry/opentelemetry-specification: specification/compatibility/opencensus.md

# OpenCensus Compatibility

## Abstract

The OpenTelemetry project aims to provide backwards compatibility with the

> [OpenCensus](https://opencensus.io) project in order to ease migration of

Reviewer: I think it could help to have an example migration. Dunno, but for example something like

- Start with OC app that uses spring-boot and mongodb instrumented
- Add OTel shim and OTel SDK as dependencies, configure OTel SDK instead of OC implementation. Should still work mostly the same
- Replace spring-boot OC instrumentation with OTel instrumentation. Should still work mostly the same
- Replace MongoDB OC instrumentation with OTel instrumentation
- Drop OC

Author: That was Paraagraph 4.  Should I expand that with this?  I do want to make sure it's clear the path-of-adoption and goals.

Reviewer: Yeah paragraph 4 is a general statement, I think an example with actual library names, etc can help a lot

Author: Ok, updated.

instrumented codebases.

This functionality will be provided as a bridge layer implementing the
[OpenCensus API](https://github.com/census-instrumentation/opencensus-specs)

> using the OpenTelemetry API. This layer MUST NOT rely on implementation specific

Reviewer: Also I believe, it MUST NOT depend on the OpenCensus implementation, only the OpenCensus API.

Author: The current shims (at least for Java) actually do, and require latest version of OpenCensus.

Reviewer: Oh - I wonder why we reimplement binaryformat then instead of using the one in the implementation, thought we were avoiding that dependency. Or maybe we haven't completed stripping it out?

https://github.com/open-telemetry/opentelemetry-java/blob/228e26ad9b4e35c4bcab8de7654ea270c5447196/opencensus-shim/src/main/java/io/opentelemetry/opencensusshim/OpenTelemetryBinaryFormatImpl.java#L20

Author: It turns out we can't really strip the dependency (in the shim).   The API surface area of OpenCenuss and the extension points require us to directly tie into the SDK.

TL;DR: the requirement that an "OpenTelemetry Sandwich" works means that we need to directly touch some SDK components.

````
|-- Applicaiton - Configured OpenTelemetry SDK --------------------------------- |
    |--  Library A -> Using OpenCensus to genereate Trace A  --------- |
      |--  Library B -> Using OpenTelemetry to generate Trace B-- |
```

If we want Trace B to discover Trace A via context and denote it as a parent, then the current approach is what we came up with.

details of the OpenTelemetry SDK.

More specifically, the intention is to allow OpenCensus instrumentation to be
recorded using OpenTelemetry. This Shim Layer MUST NOT publicly expose any
upstream OpenTelemetry API.

The OpenCensus Shim and the OpenTelemetry API/SDK are expected to be consumed
simultaneously in a running service, in order to ease migration from the former
to the latter.  It is expected that application owners will begin the migration
process towards OpenTelemetry via the shim and adding new telemetry information
via OpenTelemetry.  Slowly, libraries and integrations will also migrate
towards opentelemetry until the shim is no longer necessary.

OpenCensus supports two primary types of telemetry: Traces and Stats (Metrics).
Compatibility for these is defined separately.

> The overridding philosophy for compatibility is that OpenCensus instrumented
> libraries and applications need make *no change* to their API usage in order
> to use OpenTelemetry.   All changes should be solely configuration / setup.

## Goals

> OpenTelemetry<->OpenCensus tracing compatibility has the following goals:

Reviewer: ```suggestion
OpenTelemetry<->OpenCensus compatibility has the following goals:
```

Tracing is mentioend below

1. OpenCensus has no hard dependency on OpenTelemetry
2. Minimal changes to OpenCensus for implementation
3. Easy for users to use, ideally no change to their code

Additionally, for tracing there are the following goals:

> 1. Maintain parent-child span relationship from applications and libraries

Reviewer: ```suggestion
1. Maintain parent-child span relationship between applications and libraries
```

> 2. Maintain unscoped span relationships from applications and libraries

Reviewer: What does unscoped span relationship mean?

Author: Links and link-like behavior from other features.

Reviewer: How about using the term `link` here then - we don't have this concept of scoped span elsewhere I think, some [mentions](https://github.com/open-telemetry/opentelemetry-specification/search?q=scope) are not the same as the meaning here I think.

Author: Done

   speciifcaiton of parent spans post-initialization.
2. Links added to spans after the spans are created. This is [not supported in
   OpenTelemetry](../api.md#specifying-links), therefore OpenCensus spans that
   have links added to them after creation will be mapped to OpenTelemetry spans
   without the links.
3. OpenTelemetry specifies that samplers are
   [attached to an entire Trace provider](../trace/sdk.md#sampling)
   while [OpenCensus allows custom samplers per span](https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/Sampling.md#how-can-users-control-the-sampler-that-is-used-for-sampling).
4. TraceFlags in both OpenCensus and OpenTelemetry only specify the single
   `sampled` flag ([OpenTelemetry](../trace/api.md#spancontext),
   [OpenCensus](https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/TraceConfig.md#traceparams)).
   Some OpenCensus APIs support "debug" and "defer" tracing flags in additon to
   "sampled".  In this case, the OpenCensus bridge will do its best to support
   and translate unspecified flags into the closest OpenTelemetry equivalent.

### Context Propagation

> The shim will provide an OpenCensus `PropogationComponent` implementation which

Reviewer: ```suggestion
The shim will provide an OpenCensus `PropagationComponent` implementation which
```

Reviewer: Many of these comments have been resolved without being fixed it seems, did you miss pushing a commit?

> maps OpenCenus binary and text progation to OpenTelemetry context.

Reviewer: ```suggestion
maps OpenCenus binary and text propagation to OpenTelemetry context.
```

#### Text Context

> This adapter MUST use an OpenTelemetry `TextMapPropogator` to implement the

Reviewer: ```suggestion
This adapter MUST use an OpenTelemetry `TextMapPropagator` to implement the
```

OpenCensus `TextFormat`.

> This adapter SHOULD use configured OpenTelemetry `TextMapPropogator` on the

Reviewer: ```suggestion
This adapter SHOULD use configured OpenTelemetry `TextMapPropagator` on the
```

OpenTelemetry `TraceProvider` for text format propagation.

> This adapter MUST provide a default `W3CTraceContextPropogator` in lieu of a

Reviewer: ```suggestion
This adapter MUST provide a default `W3CTraceContextPropagator` in lieu of a
```

> propogator defined against the `TraceProvider`.

Reviewer: ```suggestion
propagator defined against the `TraceProvider`.
```

#### B3 Context

> This adapter SHOULD use an contributed OpenTelemetry `B3Propogator` for the

Reviewer: ```suggestion
This adapter SHOULD use n contributed OpenTelemetry `B3Propagator` for the
```

B3 text format.

#### OpenCensus Binary Context

> This adapter MUST implement OpenCensus `BinaryPropagator` to write OpenCensus

Reviewer: So based on https://github.com/open-telemetry/opentelemetry-specification/pull/1332/files#r557580446, since we do need a dependency on the opencensus coreimpl anyways, is there any reason to require an implementation in the adapter then? May as well save some work if this isn't really providing a benefit (complete decoupling from coreimpl).

Author: Two reasons, both technical and not ideal:

1. The utilities for writing the binary format are package-private, we can't get em (at least in Java).
2. We DO need to override it for TextPropagater.

I'm ok adjusting the specification however we see fit here.

Reviewer: Sent https://github.com/open-telemetry/opentelemetry-java/pull/2605 to demonstrate not copying the binary format implementation.

This may not be possible in all languages, but I think this could be worded to `MUST provide an implementation of ... This MAY be the same implementation as OpenCensus if it is available.`

Author: done.   Thanks for catching that, I had missed the class was not final.

> This adapter MUST implement OpenCensus `BinaryPropogator` to write OpenCensus

Reviewer: ```suggestion
This adapter MUST implement OpenCensus `BinaryPropagator` to write OpenCensus
```

# OpenCensus Compatibility

## Abstract

The OpenTelemetry project aims to provide backwards compatibility with the
[OpenCensus](https://opencensus.io) project in order to ease migration of
instrumented codebases.

This functionality will be provided as a bridge layer implementing the
[OpenCensus API](https://github.com/census-instrumentation/opencensus-specs)
using the OpenTelemetry API. This layer MUST NOT rely on implementation specific
details of the OpenTelemetry SDK.

More specifically, the intention is to allow OpenCensus instrumentation to be
recorded using OpenTelemetry. This Shim Layer MUST NOT publicly expose any
upstream OpenTelemetry API.

The OpenCensus Shim and the OpenTelemetry API/SDK are expected to be consumed
simultaneously in a running service, in order to ease migration from the former
to the latter.  It is expected that application owners will begin the migration
process towards OpenTelemetry via the shim and adding new telemetry information
via OpenTelemetry.  Slowly, libraries and integrations will also migrate
towards opentelemetry until the shim is no longer necessary.

> For example, an applciation may have traces today of the following variety:

Reviewer: ```suggestion
For example, an application may have traces today of the following variety:
```

```
|-- Applicaiton - Configured OpenCenus --------------------------------- |

>     |--  gRPC -> Using OpenCensus to genereate Trace A  --------- |

Reviewer: ```suggestion
    |--  gRPC -> Using OpenCensus to generate Trace A  --------- |
```

      |--  Application -> Using OpenCensus to generate a sub Trace B-- |
```

In this case, the application should be able to update its outer layer to
OpenTelemetry, without having to wait for all downstream dependencies to
have updated to OpenTelemetry (or deal with incompatibilities therein). The
Application also doesn't need to rewrite any of its own instrumentation.

```

> |-- Applicaiton - Configured Otel w/ OpenCensus Shim ------------------- |

Reviewer: ```suggestion
|-- Application - Configured Otel w/ OpenCensus Shim ------------------- |
```

>     |--  gRPC -> Using OpenCensus to genereate Trace A  --------- |

Reviewer: ```suggestion
    |--  gRPC -> Using OpenCensus to generate Trace A  --------- |
```

      |--  Application -> Using OpenCensus to generate a sub Trace B-- |
```

Next, the application can update its own instrumentation in a piecemeal fashion:

```

> |-- Applicaiton - Configured Otel w/ OpenCensus Shim ---------------------- |

Reviewer: ```suggestion
|-- Application - Configured Otel w/ OpenCensus Shim ---------------------- |
```

>     |--  gRPC -> Using OpenCensus to genereate Trace A  --------- |

Reviewer: ```suggestion
    |--  gRPC -> Using OpenCensus to generate Trace A  --------- |
```

      |--  Application -> Using OpenTelemetry to generate a sub Trace B-- |
```

> This layer of Otel -> OpenCensus -> Otel tracing can be thought of as the
> "OpenTelemetry sandwhich" problem, and is the key motivating factor for
> this specification.

> Finally, the APplication would update all usages of OpenCensus to OpenTelemetry.

Reviewer: ```suggestion
Finally, the Application would update all usages of OpenCensus to OpenTelemetry.
```

```

> |-- Applicaiton - Configured Otel standalong ----------------------------- |

Reviewer: ```suggestion
|-- Application - Configured Otel standalone ----------------------------- |
```

>     |--  gRPC -> Using Otel to genereate Trace A  --------- |

Reviewer: ```suggestion
    |--  gRPC -> Using Otel to generate Trace A  --------- |
```

This component MUST be an optional dependency.

### Creating Spans in OpenCensus

When the shim is in place, all OpenCensus Spans MUST be sent through an
OpenTelemetry `Tracer` as specified for the OpenTelemetry API.

This mechanism SHOULD be seamless to the user in languages that allow discovery
and auto injection of dependencies.

### Methods on Spans

All specified methods in OpenCensus will delegate to the underlying `Span` of
OpenTelemetry.

#### Known Incompatibilities

Below are listed known incompatibilities between OpenTelemetry and OpenCensus
specifications.   Applications leveraging unspecified behavior from OpenCensus
that *is* specified incompatibly within OpenTelemetry are not eligble for
using the OpenCensus <-> OpenTelemetry bridge.

1. In OpenCensus, there is [no specification](https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/Span.md#span-creation)
   when parent spans can be specified on a child span.   OpenTelemetry specifies
   that [parent spans must be specified during span creation](../trace/api.md#span-creation).
   This leads to some issues with OpenCensus APIs that allowed flexible

>    speciifcaiton of parent spans post-initialization.

Reviewer 2: s/speciifcaiton/specification

>    speciifcaiton of parent spans post-initialization.

Reviewer 2: I don't think any implementation does that. As the title says that is "span-creation" so I think this is just a misinterpretation of the OpenCensus specs.

Author: I'd encourage you to go look up the specifics here.   I know for Java, it's allowed by the API.

Reviewer 2: Prove me wrong please, I checked Go and Java

Reviewer 2: https://github.com/census-instrumentation/opencensus-go/blob/master/trace/trace.go#L173

https://github.com/census-instrumentation/opencensus-java/blob/master/api/src/main/java/io/opencensus/trace/Tracer.java#L307

Author: Right, it was Typescript.  Sorry it's been a while since I wrote this: https://github.com/census-instrumentation/opencensus-node/blob/master/packages/opencensus-core/src/trace/model/span.ts#L59

The point is, it's allowable by the API and we're not going to support it.

Reviewer 2: So you can call it that only typescript has this problem.

Author: Python does too, and likely other dynamic languages.   Do we call out specific languages in other areas of the specification?

Reviewer 2: sure

2. Links added to spans after the spans are created. This is [not supported in
   OpenTelemetry](../api.md#specifying-links), therefore OpenCensus spans that
   have links added to them after creation will be mapped to OpenTelemetry spans
   without the links.
3. OpenTelemetry specifies that samplers are
   [attached to an entire Trace provider](../trace/sdk.md#sampling)
   while [OpenCensus allows custom samplers per span](https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/Sampling.md#how-can-users-control-the-sampler-that-is-used-for-sampling).
4. TraceFlags in both OpenCensus and OpenTelemetry only specify the single
   `sampled` flag ([OpenTelemetry](../trace/api.md#spancontext),
   [OpenCensus](https://github.com/census-instrumentation/opencensus-specs/blob/master/trace/TraceConfig.md#traceparams)).
   Some OpenCensus APIs support "debug" and "defer" tracing flags in additon to
   "sampled".  In this case, the OpenCensus bridge will do its best to support

>    and translate unspecified flags into the closest OpenTelemetry equivalent.

Reviewer 2: facepalm :)))

Reviewer 2: https://github.com/open-telemetry/opentelemetry-go/issues/1515

### Context Propagation

The shim will provide an OpenCensus `PropagationComponent` implementation which
maps OpenCenus binary and text propagation to OpenTelemetry context.

#### Text Context

This adapter MUST use an OpenTelemetry `TextMapPropagator` to implement the
OpenCensus `TextFormat`.

> This adapter SHOULD use configured OpenTelemetry `TextMapPropagator` on the

Reviewer: Just checking, is this something Java can't do but other languages may? For example, we have explicit extension points for B3 and W3C

https://github.com/open-telemetry/opentelemetry-java/blob/main/opencensus-shim/src/main/java/io/opentelemetry/opencensusshim/OpenTelemetryPropagationComponentImpl.java#L28

So it seems like it wouldn't make sense to use the OTel configured propagator instead of directly exposing the correct implementations.

Author: I can clarify this.  I actually wanted to specify it so that OpenCensus shim will use the globallly configured propogator (see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md#get-global-propagator).   Right now the implementation DOES NOT do this, but I'd like to fix it.  Happy to hear thoughts/.copncerns.

My main reasoning for this shift from the Java impl is because  I'd like to give OTel configuration the ability to control the impl of the propogator.  One of the "principles" here is when adopting the shim, OTEL configuration should override OC, always.

Reviewer 2: Probably you may want to state this principal somewhere.

OpenTelemetry `TraceProvider` for text format propagation.

> This adapter MUST provide a default `W3CTraceContextPropagator` in lieu of a

Reviewer: Not sure what this sentence means. Is it the same as my above point, where we actually don't want to use the configured provider? More concretely, this line of code in Java?

https://github.com/open-telemetry/opentelemetry-java/blob/main/opencensus-shim/src/main/java/io/opentelemetry/opencensusshim/OpenTelemetryPropagationComponentImpl.java#L34

Author: NO, this is related to selecting the W3C propagatoor off the global (see: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md#get-global-propagator) and if there is none available, providing a default.

I'll take a stab at rewording this, need to think about a better phrasing.

Author: Take look at the rewording, hopefully it's more clear.

# OpenCensus Compatibility

**Status**: [Experimental](../document-status.md), Unless otherwise specified.

## Abstract

The OpenTelemetry project aims to provide backwards compatibility with the
[OpenCensus](https://opencensus.io) project in order to ease migration of
instrumented codebases.

This functionality will be provided as a bridge layer implementing the
[OpenCensus API](https://github.com/census-instrumentation/opencensus-specs)
using the OpenTelemetry API. This layer MUST NOT rely on implementation specific
details of the OpenTelemetry SDK.

More specifically, the intention is to allow OpenCensus instrumentation to be

> recorded using OpenTelemetry. This Shim Layer MUST NOT publicly expose any

Reviewer 3: Might need to clarify where the shim should go - in OpenTelemetry project or OpenCensus project.

https://github.com/open-telemetry/opentelemetry-specification/blob/9047c91412d3d4b7f28b0f7346d8c5034b509849/specification/versioning-and-stability.md#a-note-on-replacing-signals

_OpenTelemetry already supports two tracing APIs: OpenTelemetry and OpenTracing. We invented a new tracing API, but continue to support the old one._

Author: I can do.  In practice, we've actually released new OpenCensus versions that expose hooks for the shim, and the shim lives on the OTEL side (as that's where you configure your pipeline).

As an FYI for you, I expect we might have to do something similar in Python/JavaScript if we run into similar issues as what we found in Go/Java.

1. OpenCensus has no hard dependency on OpenTelemetry
2. Minimal changes to OpenCensus for implementation
3. Easy for users to use, ideally no change to their code

Additionally, for tracing there are the following goals:

1. Maintain parent-child span relationship between applications and libraries
2. Maintain span link relationships between applications and libraries

## Trace

**Status**: [Experimental, Feature Freeze](../document-status.md)

OpenTelemetry will provide an OpenCensus-Trace-Shim component that can be
added as a dependency to ensure compatibility with OpenCensus.

This component MUST be an optional dependency.

### Creating Spans in OpenCensus

When the shim is in place, all OpenCensus Spans MUST be sent through an
OpenTelemetry `Tracer` as specified for the OpenTelemetry API.

This mechanism SHOULD be seamless to the user in languages that allow discovery
and auto injection of dependencies.

### Methods on Spans

All specified methods in OpenCensus will delegate to the underlying `Span` of
OpenTelemetry.

#### Known Incompatibilities

Below are listed known incompatibilities between OpenTelemetry and OpenCensus

> specifications.   Applications leveraging unspecified behavior from OpenCensus

Reviewer 3: ```suggestion
specifications.  Applications leveraging unspecified behavior from OpenCensus
```

Reviewer 3: There are other places in the doc that also have 3 SPCs.
